### ISSUE TITLE
Trouble making HTML documentation

### ISSUE THREAD

**Issue by jw-allen** [frankluebeck/GAPDoc#44]
(If I am asking for help in the wrong place, please direct me to the right place.)

I am developing a GAP package "SBStrips" and having trouble making the HTML documentation using GAPDoc. To recreate my problem and hopefully help, you should be able to install the current version of my package from
https://github.com/jw-allen/sbstrips/releases/tag/HTMLhelp
and unpacking it in the familiar way; see also https://github.com/jw-allen/sbstrips/blob/master/README.

I have a very minimal main XML document (`doc/main.xml`) which I feed to GAP using the instructions from Chapter 5 of the GAPDoc manual. Initially, this works fine.

    path := Directory( "Documents/GitHub/pkg/sbstrips/doc/" );
    main := "main.xml";
    files := [ "../lib/1reg.gd", "../lib/overquiver.gd", "../lib/panels.gd",
     "../lib/patchreps.gd", "../lib/permdata.gd", "../lib/strips.gd",
     "../lib/syllables.gd", "../lib/util.gd", "../lib/vertseqs.gd",
     "titlepage.xml", "ch_discrete_model.xml", "ch_example.xml",
     "ch_introduction.xml", "ch_patches.xml", "ch_permissible_data.xml",
     "ch_quiver_utilities_overquiver.xml", "ch_strips_syzygies.xml",
     "ch_syllables.xml", "ch_utilities.xml", "ch_vis_encodings.xml",
     "ap_example_sbas.xml",
     "sbstripsbib.xml"
     ];
    bookname := "manual";

    # Compose document
    doc := ComposedDocument( "GAPDoc", path, main, files, true );
    
    # Make and check tree 
    r := ParseTreeXMLString( doc[1], doc[2] );
    CheckAndCleanGapDocTree( r );
    
    # Make and print text files
    t := GAPDoc2Text( r , path );
    GAPDoc2TextPrintTextFiles( t , path );
    
    # Make and print LaTeX files
    l := GAPDoc2LaTeX( r );
    FileString( Filename( path, Concatenation( bookname, ".tex") ), l );

With the above, I create the text and LaTeX files without problem (and from them, the PDF and auxiliary files). With the following, I sort out the Six files.

    AddPageNumbersToSix( r, Filename( path, Concatenation( bookname, ".pnr" ) ) );
    PrintSixFile( Filename( path, "manual.six" ), r, bookname );

So far, so good. But then I try to make the HTML files

    # Make and print HTML files
    h := GAPDoc2HTML( r, path );

This causes the following error

    #I First run, collecting cross references, index, toc, \
    bib and so on . . .
    Error, List Element: <list>[1] must have an assigned value in
      Append( str, Concatenation( "<a href=\"chapBib",
       GAPDoc2HTMLProcs.ext, "#biB", key, "\">[",
       r.root.biblabels[pos] )
     ); at /proc/cygdrive/C/Users/Joe/gap4r8/pkg/GAPDoc-1.6\
    .1/lib/GAPDoc2HTML.gi:1496 called from
    GAPDoc2HTMLProcs.(r.name)( r, str
     ); at /proc/cygdrive/C/Users/Joe/gap4r8/pkg/GAPDoc-1.6\
    .1/lib/GAPDoc2HTML.gi:529 called from
    GAPDoc2HTML( a, s
     ); at /proc/cygdrive/C/Users/Joe/gap4r8/pkg/GAPDoc-1.6\
    .1/lib/GAPDoc2HTML.gi:588 called from
    GAPDoc2HTMLContent( r, par
     ); at /proc/cygdrive/C/Users/Joe/gap4r8/pkg/GAPDoc-1.6\
    .1/lib/GAPDoc2HTML.gi:1103 called from
    GAPDoc2HTMLProcs.ChapSect( r, par, "Chapter"
     ); at /proc/cygdrive/C/Users/Joe/gap4r8/pkg/GAPDoc-1.6\
    .1/lib/GAPDoc2HTML.gi:1142 called from
    GAPDoc2HTMLProcs.(r.name)( r, str
     ); at /proc/cygdrive/C/Users/Joe/gap4r8/pkg/GAPDoc-1.6\
    .1/lib/GAPDoc2HTML.gi:529 called from
    GAPDoc2HTML( a, l
     ); at /proc/cygdrive/C/Users/Joe/gap4r8/pkg/GAPDoc-1.6\
    .1/lib/GAPDoc2HTML.gi:585 called from
    GAPDoc2HTMLProcs.(r.name)( r, str
     ); at /proc/cygdrive/C/Users/Joe/gap4r8/pkg/GAPDoc-1.6\
    .1/lib/GAPDoc2HTML.gi:529 called from
    GAPDoc2HTML( a, l
     ); at /proc/cygdrive/C/Users/Joe/gap4r8/pkg/GAPDoc-1.6\
    .1/lib/GAPDoc2HTML.gi:585 called from
    GAPDoc2HTMLContent( r, par
     ); at /proc/cygdrive/C/Users/Joe/gap4r8/pkg/GAPDoc-1.6\
    .1/lib/GAPDoc2HTML.gi:814 called from
    GAPDoc2HTMLProcs.Book( r.content[i], [  ], pi
     ); at /proc/cygdrive/C/Users/Joe/gap4r8/pkg/GAPDoc-1.6\
    .1/lib/GAPDoc2HTML.gi:655 called from
    GAPDoc2HTMLProcs.(r.name)( r, str
     ); at /proc/cygdrive/C/Users/Joe/gap4r8/pkg/GAPDoc-1.6\
    .1/lib/GAPDoc2HTML.gi:529 called from
    <function "GAPDoc2HTML">( <arguments> )
     called from read-eval loop at line 1 of *errin*

I don't know how to overcome this error?

I've poked around a lot. Having picked through the GAPDoc source code, I notice there are many references to components of `r.attributes` (where `r` is the GAPDoc tree). But, for me, `r.attributes` is the empty record.

    gap> r.attributes;
    rec(  )

Perhaps this is relevant? Regardless, thank you for reading and for any insight you can offer :)

**Comment by fingolfin** [comment]
Shameless plug: you might want to try using [AutoDoc](https://github.com/gap-packages/AutoDoc) to setup your package documentation; it also uses GAPDoc, but takes care of a bunch of the busy work for simply building the manual (and it can also do things like generating the title page of the manual for you from the data in `PackageInfo.g`); this is useful even if one doesn't use any of its other features.

(By the way, is there a reason you are targeting the rather old GAP version 4.5? Do you actually test that your package works on that?)

**Comment by frankluebeck** [comment]
I cannot reproduce the problem with any version of GAP/GAPDoc I have at hand (I tried stoneage GAP 4.4.5, GAP 4.10.2 , current release version 4.11.0 and current master branch). Your input just runs without problem and produces all output versions of the manual.
Can you give information about the GAP/GAPDoc versions you are using. Did you try using  the current GAP release?

BTW: Why don't you use the compound command `MakeGAPDocDoc` provided by GAPDoc? For example, you could call the following GAP file (put and call it in the top directory of your package):

```
##  makedocrel.g
##  
##  this creates the documentation, needs: GAPDoc package, latex, pdflatex,
##  mkindex

# substitute this by path to main GAP directory, if this package is not
# in standard location
#pathtoroot := ...
if IsBound(pathtoroot) then
  relpath := pathtoroot;
else
  relpath:="../../..";
fi;
LoadPackage("GAPDoc");

path := "doc";
main := "main.xml";
files := [ "../lib/1reg.gd", "../lib/overquiver.gd", "../lib/panels.gd",
     "../lib/patchreps.gd", "../lib/permdata.gd", "../lib/strips.gd",
     "../lib/syllables.gd", "../lib/util.gd", "../lib/vertseqs.gd",
     "titlepage.xml", "ch_discrete_model.xml", "ch_example.xml",
     "ch_introduction.xml", "ch_patches.xml", "ch_permissible_data.xml",
     "ch_quiver_utilities_overquiver.xml", "ch_strips_syzygies.xml",
     "ch_syllables.xml", "ch_utilities.xml", "ch_vis_encodings.xml",
     "ap_example_sbas.xml",
     "sbstripsbib.xml" ];
bookname := "SBStrips";

MakeGAPDocDoc(path, main, files, bookname, relpath, "MathJax");

# make HTML versions look nicer
CopyHTMLStyleFiles("doc");

# compatibility with old style package manuals
GAPDocManualLab("SBStrips");

QUIT;
```

Except for the second last line (which causes to load your package) this also works for me with all versions of GAP/GAPDoc mentioned above.

**Comment by jw-allen** [comment]
**[TLDR: Issue solved, all stupid errors on my part, thanks to you both!]**

> (By the way, is there a reason you are targeting the rather old GAP version 4.5? Do you actually test that your package works on that?)

> Can you give information about the GAP/GAPDoc versions you are using. Did you try using the current GAP release?

I was previously using GAP 4.8.x, featuring GAPDoc 1.6.y (I think). This is simply because I first downloaded GAP some time ago and did not think to update -- more fool me. I have updated to the current version 4.11.0 of GAP, which features GAPDoc 1.6.3. The update has fixed my problem. (Having uninstalled the old version, I forget the exact values of "x" and "y".)

> BTW: Why don't you use the compound command `MakeGAPDocDoc` provided by GAPDoc?

Once again, the real error here is mine. When I started trying to document my package, I could not get `MakeGAPDocDoc` to work. I forget the precise reason. Through a combination of panic/lack of confidence (I have no formal background in coding/programming) and some laziness (the GAPDoc documentation says `MakeGAPDocDoc`'s behaviour is system-dependent and can require tinkering, which I did not feel up to), I settled for a partial solution. Namely, performing the constituent commands of `MakeGAPDocDoc` manually.

> For example, you could call the following GAP file

Thank you very much for this. I have included/will include it in my package, with the attribution due to you. It works splendidly. In fact, I've run it a couple of times in a row out of sheer happiness.

> Shameless plug

Thank you for the recommendation. I think I tried AutoDoc before, immediately encountered some (probably very basic/silly) problems, panicked and then abandoned AutoDoc, thinking I would just "cut out the middleman" and make my documentation directly with GAPDoc instead. I will give your package another go and see how I get on.

Thank you both again!
